src/Authentication/ApiKey.php

```diff
@@ -0,0 +1,379 @@
+	 *
+	 * @var string
+	 */
+	const META_PREFIX = 'satispress_api_key.';
+
+	/**
+	 * API key length.
+	 *
+	 * @var integer
+	 */
+	const TOKEN_LENGTH = 32;
+
+	/**
+	 * String to prepend to API keys.
+	 *
+	 * @var string
+	 */
+	const TOKEN_PREFIX = '';
+
+	/**
+	 * API key data.
+	 *
+	 * @var array
+	 */
+	protected $data;
+
+	/**
+	 * API key token.
+	 *
+	 * @var string
+	 */
+	protected $token;
+
+	/**
+	 * User associated with the API key.
+	 *
+	 * @var WP_User
+	 */
+	protected $user;
+
+	/**
+	 * Initialize an API key.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @param WP_User $user  WordPress user.
+	 * @param string  $token API key token.
+	 * @param array   $data  Additional data associated with the key.
+	 */
+	public function __construct( WP_User $user, string $token, array $data = [] ) {
+		$this->user  = $user;
+		$this->token = $token;
+		$this->data  = $data;
+	}
+
+	/**
+	 * Retrieve the API Key name.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return string
+	 */
+	public function get_name(): string {
+		return $this->data['name'] ?? '';
+	}
+
+	/**
+	 * Retrieve the API Key token.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return string
+	 */
+	public function get_token(): string {
+		return $this->token;
+	}
+
+	/**
+	 * Retrieve the user associated with the API Key.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return WP_User
+	 */
+	public function get_user(): WP_User {
+		return $this->user;
+	}
+
+	/**
+	 * Retrieve and format a date field.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @param string $name   Field name.
+	 * @param string $format Optional. Date format.
+	 * @return mixed
+	 */
+	public function get_date( string $name, string $format = null ) {
+		if ( empty( $this->data[ $name ] ) ) {
+			return '';
+		}
+
+		return $this->format_date( $this->data[ $name ], $format );
+	}
+
+	/**
+	 * Revoke the API key.
+	 *
+	 * @since 0.3.0
+	 */
+	public function revoke() {
+		delete_user_meta(
+			$this->get_user()->ID,
+			$this->get_meta_key( $this->token )
+		);
+	}
+
+	/**
+	 * Save the API Key.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return $this
+	 */
+	public function save(): self {
+		update_user_meta(
+			$this->get_user()->ID,
+			static::get_meta_key( $this->token ),
+			$this->data
+		);
+
+		return $this;
+	}
+
+	/**
+	 * Convert the API Key to an array.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @return array
+	 */
+	public function to_array(): array {
+		return [
+			'name'      => $this['name'] ?? '',
+			'user'      => $this->get_user()->ID,
+			'token'     => $this->get_token(),
+			'last_used' => $this->get_date( 'last_used' ),
+			'created'   => $this->get_date( 'created' ),
+		];
+	}
+
+	/**
+	 * Create and save a new API key for a user.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @param WP_User $user WordPress user.
+	 * @param array   $data Optional. Additional data associated with the API key.
+	 * @return ApiKey
+	 */
+	public static function create( WP_User $user, array $data = [] ): ApiKey {
+		if ( ! isset( $data['created'] ) ) {
+			$data['created'] = time();
```

Comment:
This is an external dependency (system time) - see https://matthiasnoback.nl/2018/02/mocking-at-architectural-boundaries-persistence-and-time/ for how to handle this so that it can be mocked for testing.

Reply:
I get the why behind that, but it feels like overkill. If `time()` was used more often, that would make perfect sense, but why not pull it out into a method that can be stubbed in tests?